sql/src/main/java/io/crate/execution/jobs/JobExecutionContext.java

```diff
@@ -207,14 +248,42 @@ public long kill() {
         return numKilled;
     }
 
-    private void finish() {
+    private void close() {
         if (failure != null) {
             finishedFuture.completeExceptionally(failure);
         } else {
             finishedFuture.complete(null);
         }
     }
 
+    private void finish() {
+        if (profiler.enabled()) {
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace("Profiling is enabled. JobExecutionContext will not be closed until results are collected!");
+                LOGGER.trace("Profiling results for job {}: {}", jobId, profiler.getAsMap());
+            }
+            assert profilingFuture != null : "profilingFuture must not be null";
+            profilingFuture.complete(executionTimes());
+        } else {
+            close();
+        }
+    }
+
+    public CompletableFuture<Map<String, Long>> finishProfiling() {
+        if (profiler.enabled()) {
+            assert profilingFuture != null : "profilingFuture must not be null";
+            return profilingFuture.whenComplete((o, t) -> close());
+        }
+        // sanity check
+        IllegalStateException stateException = new IllegalStateException(String.format(Locale.ENGLISH, "Tried to finish profiling job [id=%s], but profiling is not enabled.", jobId));
+        return CompletableFutures.failedFuture(stateException);
```

Comment:
I would write this the other way around, first handling the error case, then the regular case. When I see an if condition at the beginning of a function, I assume it handles a special case rather than the expected case.